aten/src/ATen/native/cpu/layer_norm_kernel.cpp

```diff
@@ -94,60 +107,159 @@ void LayerNormBackwardKernelImplInternal(
     Tensor* dX,
     Tensor* dgamma,
     Tensor* dbeta) {
+  using Vec = vec256::Vec256<T>;
   DCHECK_EQ(dY.numel(), M * N);
   DCHECK_EQ(X.numel(), M * N);
   DCHECK_EQ(mean.numel(), M);
   DCHECK_EQ(rstd.numel(), M);
   DCHECK(!gamma.defined() || gamma.numel() == N);
-  const T* dY_data = dY.template data_ptr<T>();
-  const T* X_data = X.template data_ptr<T>();
-  const T* mean_data = mean.template data_ptr<T>();
-  const T* rstd_data = rstd.template data_ptr<T>();
-  const T* gamma_data =
-      gamma.defined() ? gamma.template data_ptr<T>() : nullptr;
+  T* dY_data = dY.template data_ptr<T>();
+  T* X_data = X.template data_ptr<T>();
+  T* mean_data = mean.template data_ptr<T>();
+  T* rstd_data = rstd.template data_ptr<T>();
+  T* gamma_data = gamma.defined() ? gamma.template data_ptr<T>() : nullptr;
   T* dX_data = dX->defined() ? dX->template data_ptr<T>() : nullptr;
   T* dgamma_data = dgamma->defined() ? dgamma->template data_ptr<T>() : nullptr;
-  if (dgamma_data != nullptr) {
-    std::memset(dgamma_data, 0, N * sizeof(T));
-  }
   T* dbeta_data = dbeta->defined() ? dbeta->template data_ptr<T>() : nullptr;
-  if (dbeta_data != nullptr) {
-    std::memset(dbeta_data, 0, N * sizeof(T));
-  }
   const T scale = T(1) / static_cast<T>(N);
   const bool gamma_null = gamma_data == nullptr;
-  for (int64_t i = 0; i < M; ++i) {
-    const T* dY_ptr = dY_data + i * N;
-    const T* X_ptr = X_data + i * N;
-    if (dX_data != nullptr) {
-      T* dX_ptr = dX_data + i * N;
-      T ds = 0;
-      T db = 0;
-      for (int64_t j = 0; j < N; ++j) {
-        const T gamma_v = gamma_null ? T(1) : gamma_data[j];
-        ds += dY_ptr[j] * X_ptr[j] * gamma_v;
-        db += dY_ptr[j] * gamma_v;
+  const bool dX_null = dX_data == nullptr;
+  const bool dgamma_null = dgamma_data == nullptr;
+  const bool dbeta_null = dbeta_data == nullptr;
+
+  // 1. Use two path parallel reduction for dgamma and dbeta:
+  //    First path: allocate an immediate buffer of size {2, max_threads, N},
+  //        dgamma_buffer = buffer[0], dbeta_buffer = buffer[1]
+  //    Parallel along dim0 and reduce dY and X along dim0 to buffer.
+  //    Second path: parallel along dim1 and reduce buffer to dgamma and dbeta.
+  //
+  // 2. Fuse first path of dgamma/dbeta with dX to reuse X[i] and dY[i] in L1 cache.
+  //
+  int num_threads = at::get_num_threads();
```

Comment:
@VitalyFedyunin Correct me if I am wrong.
It seems to me that at::get_num_threads() does not always return the same value, this might depend on the current condition of the hardware (I am not sure about this).

If we resize_ buffer here, and use this buffer in the following at::parallel_for, will this cause problem due to potential different returned value from at::get_num_threads() in at::parallel_for()?

Reply:
In theory, it might be another user thread which can potentially change max threads number between `resize_` and `parallel_for` call. For maximum safety, I recommend asserting on  `tid <  num_threads`

Reply:
Added assert for thread id check!

I did not keep `mask` here to record which `tid` has been used. The immediate buffer is zeroed, so won't matter here.

This piece of code is a copy/paste from onednn: https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/simple_layer_normalization.cpp#L204

Except that i fuse the first two loops together, this will save memory bandwidth for `dY` and `X`.